constraint/pkg/client/drivers/rego/builtin.go

```diff
@@ -25,12 +31,78 @@ func externalDataBuiltin(d *Driver) func(bctx rego.BuiltinContext, regorequest *
 			return externaldata.HandleError(http.StatusBadRequest, err)
 		}
 
-		externaldataResponse, statusCode, err := d.sendRequestToProvider(bctx.Context, &provider, regoReq.Keys, clientCert)
-		if err != nil {
-			return externaldata.HandleError(statusCode, err)
+		// check provider response cache
+		var providerRequestKeys []string
+		var providerResponseStatusCode int
+		var prepareResponse externaldata.Response
+
+		prepareResponse.Idempotent = true
+		for _, k := range regoReq.Keys {
+			cachedResponse, err := d.providerResponseCache.Get(
+				externaldata.CacheKey{
+					ProviderName: regoReq.ProviderName,
+					Key:          k,
+				},
+			)
+			if err != nil || time.Since(time.Unix(cachedResponse.Received, 0)) > d.providerResponseCache.TTL {
+				// key is not found or cache entry is stale, add key to the provider request keys
+				providerRequestKeys = append(providerRequestKeys, k)
+			} else {
+				prepareResponse.Items = append(
+					prepareResponse.Items, externaldata.Item{
+						Key:   k,
+						Value: cachedResponse.Value,
+						Error: cachedResponse.Error,
+					},
+				)
+
+				// we are taking conservative approach here, if any of the cached response is not idempotent
+				// we will mark the whole response as not idempotent
+				if !cachedResponse.Idempotent && prepareResponse.Idempotent {
+					prepareResponse.Idempotent = false
+				}
+			}
+		}
+
+		if len(providerRequestKeys) > 0 {
+			externaldataResponse, statusCode, err := d.sendRequestToProvider(bctx.Context, &provider, providerRequestKeys, clientCert)
+			if err != nil {
+				return externaldata.HandleError(statusCode, err)
+			}
+
+			for _, item := range externaldataResponse.Response.Items {
+				d.providerResponseCache.Upsert(
+					externaldata.CacheKey{
+						ProviderName: regoReq.ProviderName,
+						Key:          item.Key,
+					},
+					externaldata.CacheValue{
+						Received:   time.Now().Unix(),
+						Value:      item.Value,
+						Error:      item.Error,
+						Idempotent: externaldataResponse.Response.Idempotent,
+					},
+				)
+
+				// we are taking conservative approach here, if any of the response is not idempotent
+				// we will mark the whole response as not idempotent
+				if !externaldataResponse.Response.Idempotent && prepareResponse.Idempotent {
```

Comment:
This should be outside of the loop, since it is only set once in the response object. Also, no need for the second part of the if statement.